corehq/util/couch.py

```diff
@@ -33,3 +36,130 @@ def categorize_bulk_save_errors(error):
         result_map[error].append(result)
 
     return result_map
+
+
+class IterDB(object):
+    """
+    Bulk save docs in chunks.
+
+        with IterDB(db) as iter_db:
+            for doc in iter_docs(db, ids):
+                iter_db.save(doc)
+
+        iter_db.error_ids  # docs that errored
+        iter_db.saved_ids  # docs that saved correctly
+    """
+    def __init__(self, database, chunksize=100, throttle_secs=None):
+        self.db = database
+        self.chunksize = chunksize
+        self.throttle_secs = throttle_secs
+        self.saved_ids = []
+        self.deleted_ids = []
+        self.error_ids = []
+
+    def __enter__(self):
+        self.to_save = []
+        self.to_delete = []
+        return self
+
+    def commit(self, cmd, docs):
+        try:
+            results = cmd(docs)
+        except BulkSaveError as e:
+            results = e.results
+            error_ids = {d['id'] for d in e.errors}
+            self.error_ids.extend(error_ids)
+            success_ids = [d['id'] for d in results
+                           if d['id'] not in error_ids]
+        else:
+            success_ids = [d['id'] for d in results]
+
+        if self.throttle_secs:
+            sleep(self.throttle_secs)
+        return success_ids
+
+    def commit_save(self):
+        success_ids = self.commit(self.db.bulk_save, self.to_save)
+        self.saved_ids.extend(success_ids)
+        self.to_save = []
+
+    def commit_delete(self):
+        success_ids = self.commit(self.db.bulk_delete, self.to_delete)
+        self.deleted_ids.extend(success_ids)
+        self.to_delete = []
+
+    def save(self, doc):
+        self.to_save.append(doc)
+        if len(self.to_save) >= self.chunksize:
+            self.commit_save()
+
+    def delete(self, doc):
+        self.to_delete.append(doc)
+        if len(self.to_delete) >= self.chunksize:
+            self.commit_delete()
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        if self.to_save:
+            self.commit_save()
+        if self.to_delete:
+            self.commit_delete()
+
+
+def iter_update(db, fn, ids):
```

Comment:
I'm not opposed to that, I just don't expect that it'll actually get any use, so I figured I'd let someone add that in if they want it.